Framework/Algorithms/src/Stitch1D.cpp

```diff
@@ -1,45 +1,95 @@
 #include "MantidKernel/ArrayProperty.h"
+#include "MantidKernel/BoundedValidator.h"
+#include "MantidKernel/MultiThreaded.h"
 #include "MantidKernel/PropertyWithValue.h"
 #include "MantidKernel/RebinParamsValidator.h"
-#include "MantidKernel/MultiThreaded.h"
-#include "MantidKernel/BoundedValidator.h"
 
+#include <algorithm>
 #include <boost/tuple/tuple.hpp>
 #include <boost/format.hpp>
-#include <boost/algorithm/string.hpp>
 #include <boost/math/special_functions.hpp>
-#include <algorithm>
+#include <map>
 
-using namespace Mantid::Kernel;
 using namespace Mantid::API;
-using Mantid::HistogramData::HistogramY;
+using namespace Mantid::Kernel;
 using Mantid::HistogramData::HistogramE;
+using Mantid::HistogramData::HistogramDx;
+using Mantid::HistogramData::HistogramX;
+using Mantid::HistogramData::HistogramY;
+using Mantid::HistogramData::Points;
 
 namespace {
-
+/// Returns a tuple holding the first and last x value of the first spectrum and
+/// the lhs and rhs workspace, respectively
 using MinMaxTuple = boost::tuple<double, double>;
-MinMaxTuple calculateXIntersection(MatrixWorkspace_sptr lhsWS,
-                                   MatrixWorkspace_sptr rhsWS) {
+MinMaxTuple calculateXIntersection(MatrixWorkspace_const_sptr &lhsWS,
+                                   MatrixWorkspace_const_sptr &rhsWS) {
   return MinMaxTuple(rhsWS->x(0).front(), lhsWS->x(0).back());
 }
 
+/// Check if a double is not zero and returns a bool indicating success
 bool isNonzero(double i) { return (0 != i); }
+
+/** Sort x axis
+ @param ws :: Input workspace
+ @return A shared pointer to the resulting MatrixWorkspace
+ */
+void sortXAxis(MatrixWorkspace_sptr &ws) {
+  // using a std::multimap (keys are sorted)
+  std::multimap<double, double> sorter;
+  double x_value;
+  const int nHists = static_cast<int>(ws->getNumberHistograms());
+  for (int i = 0; i < nHists; ++i) {
+    for (size_t k = 0; k < ws->size(); ++k) {
+      x_value = ws->x(i)[k];
+      sorter.insert(std::pair<double, double>(x_value, ws->y(i)[k]));
+      sorter.insert(std::pair<double, double>(x_value, ws->e(i)[k]));
+      if (ws->hasDx(i) && !ws->isHistogramData())
+        sorter.insert(std::pair<double, double>(x_value, ws->dx(i)[k]));
+    }
+    size_t ws_size = ws->size();
+    Mantid::MantidVec vecx(ws_size);
+    Mantid::MantidVec vecy(ws_size);
+    Mantid::MantidVec vece(ws_size);
+    Mantid::MantidVec vecdx(ws_size);
+    int l = 0;
+    auto it = sorter.cbegin();
+    while (it != sorter.cend()) {
+      vecx[l] = it->first;
+      vecy[l] = it->second;
+      vece[l] = (++it)->second;
+      if (ws->hasDx(i) && !ws->isHistogramData())
+        vecdx[l] = (++it)->second;
+      ++l;
+      ++it;
+    }
+    auto x = make_cow<HistogramX>(std::move(vecx));
+    auto y = make_cow<HistogramY>(std::move(vecy));
+    auto e = make_cow<HistogramE>(std::move(vece));
+    ws->setSharedX(i, x);
+    ws->setSharedY(i, y);
+    ws->setSharedE(i, e);
```

Comment:
Using `make_cow` and `setSharedXYE` is not helping here since it is not reused for different workspace indices, i.e., there will not be any sharing. I would suggest to directly use the accessors like `mutableX`` etc., instead of using temporary vectors `vecx` etc., avoiding allocation overheads.

Reply:
Sounds reasonable.